File: frontend/src/engine/capabilities/writeFormula.ts
/**
 * writeFormula – Write a formula to a cell, optionally fill down.
 *
 * This is the preferred action when the planner can express the operation
 * as a native Excel formula. Native formulas are better because:
 * - They recalculate automatically when source data changes
 * - They are visible and auditable by the user
 * - They leverage Excel's optimized calculation engine
 *
 * Office.js notes:
 * - range.formulas accepts a 2D string array where each string starts with "=".
 * - fillDown() copies the formula and adjusts relative references automatically.
 * - For XLOOKUP, FILTER, and other dynamic array formulas, the formula
 *   "spills" into adjacent cells automatically (Excel 365+).
 */

import { CapabilityMeta, WriteFormulaParams, StepResult, ExecutionOptions } from "../types";
import { registry } from "../capabilityRegistry";
import { resolveRange } from "./rangeUtils";

const meta: CapabilityMeta = {
  action: "writeFormula",
  description: "Write a formula to a cell, optionally fill down",
  mutates: true,
  affectsFormatting: false,
};

async function handler(
  context: Excel.RequestContext,
  params: WriteFormulaParams,
  options: ExecutionOptions
): Promise<StepResult> {
  const { cell, formula, fillDown } = params;

  if (options.dryRun) {
    return {
      stepId: "",
      status: "success",
      message: `Would write formula ${formula} to ${cell}${fillDown ? ` and fill down ${fillDown} rows` : ""}`,
    };
  }

  options.onProgress?.(`Writing formula to ${cell}...`);

  const range = resolveRange(context, cell);

  // Write the formula to the single cell
  range.formulas = [[formula]];
  await context.sync();

  // Optionally fill down — autoFill adjusts relative references automatically
  if (fillDown && fillDown > 1) {
    options.onProgress?.(`Filling formula down ${fillDown} rows...`);
    const fillRange = range.getResizedRange(fillDown - 1, 0);
    range.autoFill(fillRange, Excel.AutoFillType.fillDefault);
    await context.sync();
  }

  // Check for formula errors after writing (#SPILL!, #REF!, #VALUE!, #NAME?, etc.)
  const checkRange = fillDown && fillDown > 1
    ? range.getResizedRange(Math.min(fillDown - 1, 5), 0) // check first few rows
    : range;
  checkRange.load("values");
  await context.sync();

  const errorTypes = ["#SPILL!", "#REF!", "#VALUE!", "#NAME?", "#NULL!", "#N/A", "#DIV/0!", "#CALC!"];
  const firstVal = String(checkRange.values?.[0]?.[0] ?? "");
  const hasError = errorTypes.some((e) => firstVal.includes(e));

  const rowInfo = fillDown && fillDown > 1 ? ` and filled down ${fillDown} rows` : "";
  if (hasError) {
    return {
      stepId: "",
      status: "error",
      message: `Formula wrote to ${cell}${rowInfo} but produced ${firstVal}. The formula may need to be corrected.`,
      error: `Formula error: ${firstVal}`,
    };
  }

  return {
    stepId: "",
    status: "success",
    message: `Wrote formula to ${cell}${rowInfo}`,
  };
}



registry.register(meta, handler as any);
export { meta };

File: backend/app/execution/capabilities/write_formula.py
"""
writeFormula — write a formula to a cell, optionally fill down.

Dynamic-array functions (FILTER, UNIQUE, XLOOKUP, SORT, SEQUENCE) are
supported natively by modern desktop Excel running under xlwings, so we
don't need the Item 3 dynamic-array rewriter here. If the user's desktop
Excel happens to be 2016/2019 the formula will evaluate to #NAME? — the
post-write error check catches that and reports it back through MCP.
"""

from __future__ import annotations

from typing import Any

from app.execution.base import ExecutorContext
from app.execution.capability_registry import registry
from app.execution.range_utils import resolve_range


_ERRS = ("#SPILL!", "#REF!", "#VALUE!", "#NAME?", "#NULL!", "#N/A", "#DIV/0!", "#CALC!")


def handler(ctx: ExecutorContext, params: dict[str, Any]) -> dict[str, Any]:
    cell = params.get("cell")
    formula = params.get("formula")
    fill_down = params.get("fillDown")
    if not cell or not formula:
        return {"status": "error", "message": "writeFormula requires 'cell' and 'formula'."}

    if ctx.dry_run:
        suffix = f" and fill down {fill_down} rows" if fill_down else ""
        return {
            "status": "preview",
            "message": f"Would write formula {formula} to {cell}{suffix}.",
        }

    rng = resolve_range(ctx.workbook_handle, cell)
    rng.formula = formula

    if fill_down and fill_down > 1:
        # xlwings doesn't expose autoFill directly. Mimic Excel's AutoFill
        # semantics by copying the formula down — relative references adjust
        # automatically because we use Range.copy/Range.paste (not literal
        # string duplication).
        source = rng
        target = rng.resize(fill_down, 1)
        source.copy(target)

    # Post-write error check — read back the first cell's evaluated value.
    first_val = str(rng.value) if rng.count == 1 else str((rng.value[0] if isinstance(rng.value, list) else rng.value))
    if any(err in first_val for err in _ERRS):
        return {
            "status": "error",
            "message": f"Formula wrote to {cell} but produced {first_val}.",
            "error": f"Formula error: {first_val}",
        }

    filled = f" (filled down {fill_down} rows)" if fill_down else ""
    return {
        "status": "success",
        "message": f"Wrote formula to {rng.address}{filled}.",
        "outputs": {"cell": rng.address},
    }


registry.register("writeFormula", handler, mutates=True)
